src/core/router/transports/ssu/session.cc

```diff
@@ -716,73 +716,85 @@ void SSUSession::ProcessRelayResponse(SSUPacket* pkt) {
   context.UpdateAddress(our_IP);
 }
 
-// TODO(anonimal): bytestream refactor
 void SSUSession::SendRelayResponse(
-    std::uint32_t nonce,
+    const std::uint32_t nonce,
     const boost::asio::ip::udp::endpoint& from,
     const std::uint8_t* intro_key,
-    const boost::asio::ip::udp::endpoint& to) {
-  std::array<std::uint8_t, 80 + SSUSize::BufferMargin> buf{
-      {}};  // 64 Alice's ipv4 and 80 Alice's ipv6
-  auto payload = buf.data() + SSUSize::HeaderMin;
-  // Charlie's address always v4
-  if (!to.address().is_v4()) {
-    LOG(error)
-      << "SSUSession:" << GetFormattedSessionInfo()
-      << __func__ << ": Charlie's IP must be V4";
-    return;
-  }
-  *payload = 4;
-  payload++;  // size
-  core::OutputByteStream::Write<std::uint32_t>(
-      payload, to.address().to_v4().to_ulong());  // Charlie's IP
-  payload += 4;  // address
-  core::OutputByteStream::Write<std::uint16_t>(payload, to.port());  // Charlie's port
-  payload += 2;  // port
-  // Alice
-  auto is_IPv4 = from.address().is_v4();  // Alice's
-  if (is_IPv4) {
-    *payload = 4;
-    payload++;  // size
-    // Alice's IP V4
-    memcpy(payload, from.address().to_v4().to_bytes().data(), 4);
-    payload += 4;  // address
-  } else {
-    *payload = 16;
-    payload++;  // size
-    // Alice's IP V6
-    memcpy(payload, from.address().to_v6().to_bytes().data(), 16);
-    payload += 16;  // address
-  }
-  core::OutputByteStream::Write<std::uint16_t>(payload, from.port());  // Alice's port
-  payload += 2;  // port
-  core::OutputByteStream::Write<std::uint32_t>(payload, nonce);
-  auto relay_response = SSUPayloadType::RelayResponse;
-  if (m_State == SessionState::Established) {
-    // encrypt with session key
-    FillHeaderAndEncrypt(
-        relay_response,
-        buf.data(),
-        is_IPv4 ? 64 : 80);
-    Send(
-        buf.data(),
-        is_IPv4 ? 64 : 80);
-  } else {
-    // encrypt with Alice's intro key
-    std::array<std::uint8_t, SSUSize::IV> iv;
-    kovri::core::RandBytes(iv.data(), iv.size());
-    FillHeaderAndEncrypt(
-        relay_response,
-        buf.data(),
-        is_IPv4 ? 64 : 80,
-        intro_key,
-        iv.data(),
-        intro_key);
-    m_Server.Send(
-        buf.data(),
-        is_IPv4 ? 64 : 80,
-        from);
-  }
+    const boost::asio::ip::udp::endpoint& to)
+{
+  // Charlie's address must always be IPv4
+  if (!to.address().is_v4())
+    {
+      LOG(error) << "SSUSession:" << GetFormattedSessionInfo() << __func__
+                 << ": Charlie's must use IPv4";
+      // TODO(anonimal): don't throw?...
```

Comment:
FWIW, one obvious benefit to throwing is the caller knowing the SSU session failed. What are the downsides to throwing here?

Reply:
This will be addressed once a new message generator/dispatcher is implemented.